### ISSUE TITLE
Correct way to deal with --openssldir for bundled deployment

### ISSUE THREAD

**Issue by daniel-falk** [openssl/openssl#20644]
Due to some very specific reasons we need to build an application which bundles it's own openssl library together with certs, config files and engines. This means that we deploy a directory containing the main binary and a folder `lib` with the shared objects, among other, `libssl` and `libcrypto`. To make this application link correctly to the libraries in any directory we can of course set `RPATH` to `$ORIGIN/lib`.

The challenge we have is that during compilation of openssl we need to set the `--prefix` and `--openssldir` paths which of course is not known during compile time since these files are also bundled with the application.
```bash
./Configure shared --prefix=/library --openssldir=/lib
```
After this we can see that these paths gets hard coded into the library:
```bash
$ grep -nra /library/ ./lib/libcrypto.so
167534:/library/lib/ct_log_list.cnf
167649:OPENSSLDIR: "/library/lib"
167651:ENGINESDIR: "/library//lib/engines-1.1"
179919:/library//lib/engines-1.1
196978:/library/lib/private
196982:/library/lib
196986:/library/lib/certs
196988:/library/lib/cert.pem
```

The only way I have come up with so far that makes the library portable is to compile the library with a really long placeholder path (since the offsets in an elf-file are fixed) and then replace that string in the elf-files to the location of the application followed by trailing slashes to make the length of the strings the same before starting the application. I would hope there is a better way to this than editing the elf files.
```bash
./Configure shared --prefix=/$(printf "%0255d" 0) --openssldir=PLACEHOLDERDIR
```

**Comment by selvanair** [comment]
If the location of config, module and engine paths are known relative to that of the application, you can set the env vars OPENSSL_CONF, OPENSSL_ENGINES, OPENSSL_MODULES at application startup. That will override the built-in defaults.

**Comment by daniel-falk** [comment]
Thanks @selvanair! It's not optimal for me with env vars, but it's much better than what I was doing!

**Comment by nhorman** [comment]
marking as inactive, will be closed at the completion of 3.4 dev without further input
